features/dd-sdk-android-session-replay/src/main/kotlin/com/datadog/android/sessionreplay/internal/recorder/resources/ResourceResolver.kt

```diff
@@ -0,0 +1,328 @@
+
+                override fun onFailure() {
+                    resolveResourceCallback.onFailed()
+                }
+            }
+        )
+    }
+
+    @WorkerThread
+    @Suppress("ReturnCount")
+    private fun tryToGetBitmapFromBitmapDrawable(
+        drawable: BitmapDrawable,
+        bitmapFromDrawable: Bitmap,
+        resolveResourceCallback: ResolveResourceCallback
+    ): Bitmap? {
+        drawableUtils.createScaledBitmap(bitmapFromDrawable)?.let { scaledBitmap ->
+            val byteArray = webPImageCompression.compressBitmap(scaledBitmap)
+
+            // failed to get byteArray potentially because the bitmap was recycled before imageCompression
+            // we'll now failover to attempting to create a new bitmap from the drawable
+            if (byteArray.isEmpty() && scaledBitmap.isRecycled) {
+                return null
+            }
+
+            /**
+             * Check whether the scaled bitmap is the same as the original.
+             * Since Bitmap.createScaledBitmap will return the original bitmap if the
+             * requested dimensions match the dimensions of the original
+             */
+            val shouldCacheBitmap = scaledBitmap != drawable.bitmap
+
+            resolveResourceHash(
+                drawable = drawable,
+                bitmap = scaledBitmap,
+                byteArray = byteArray,
+                shouldCacheBitmap = shouldCacheBitmap,
+                resolveResourceCallback = resolveResourceCallback
+            )
+
+            return scaledBitmap
+        }
+
+        return null
+    }
+
+    private fun tryToGetResourceFromCache(
+        drawable: Drawable
+    ): String? = bitmapCachesManager.getFromResourceCache(drawable)
+
+    private fun shouldUseDrawableBitmap(drawable: BitmapDrawable): Boolean {
+        return drawable.bitmap != null &&
+                !drawable.bitmap.isRecycled &&
+                drawable.bitmap.width > 0 &&
+                drawable.bitmap.height > 0
+    }
+
+    // endregion
+
+    // region builder
+    internal class Builder(
+        private var applicationId: String,
+        private var recordedDataQueueHandler: DataQueueHandler = NoopDataQueueHandler(),
+        private var logger: InternalLogger = InternalLogger.UNBOUND,
+        private var threadPoolExecutor: ExecutorService = THREADPOOL_EXECUTOR,
+        private var bitmapPool: BitmapPool,
+        private var resourcesLRUCache: Cache<Drawable, CacheData>,
+        private var webPImageCompression: ImageCompression,
+        private var bitmapCachesManager: BitmapCachesManager =
+            BitmapCachesManager.Builder(
+                resourcesLRUCache,
+                bitmapPool,
+                logger
+            ).build(),
+        private var drawableUtils: DrawableUtils = DrawableUtils(
+            logger = logger,
+            bitmapCachesManager = bitmapCachesManager
+        ),
+        private var md5HashGenerator: MD5HashGenerator = MD5HashGenerator(logger)
```

Comment:
why `var` everywhere? cannot they be `val`? Also it seems we can simply use constructor directly instead of Builder?